### PR TITLE
Migrate to npm@^7.0.0, update dependencies, fix security issues

### LINKED ISSUE #1025: Upgrade packages: urix, resolve-url, requests, mkdirp-promise
I am trying to run bids-validator through npm  (bids-validator 1.5.4).  This error seems to be fixed in master, but the version on npm crashes:

bids-validator --no-config AdvancedGuardDuty
C:\Users\Kay Robbins\AppData\Roaming\npm\node_modules\bids-validator\utils\type.js:1
Error: Cannot find module '../bids_validator/rules/fixed_top_level_names.json'

### LINKED ISSUE #1106: Warnings during build
```
npm WARN deprecated urix@0.1.0: Please see https://github.com/lydell/urix#deprecated
npm WARN deprecated resolve-url@0.2.1: https://github.com/lydell/resolve-url#deprecated

npm WARN deprecated request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142

npm WARN deprecated har-validator@5.1.5: this library is no longer supported

npm WARN deprecated mkdirp-promise@5.0.1: This package is broken and no longer maintained. 'mkdirp' itself supports promises now, please switch to that.

/usr/local/bin/bids-validator -> /usr/local/lib/node_modules/bids-validator/bin/bids-validator
npm WARN @octokit/plugin-request-log@1.0.2 requires a peer of @octokit/core@>=3 but none is installed. You must install peer dependencies yourself.
```

Can these be easily resolved?

### PR DESCRIPTION
**Pull request by sappelhoff**
Trying to address some of the security concerns.

I simply navigated to `./bids-validator` and ran `npm audit fix`

Then I repeated the same in `./bids-validator-web`.

The updates are all non-breaking (because I did not use `--force`), so it should be safe to merge.

~maybe related: #1106, #1025, #1188~

closes #1106 
closes #1025 

EDIT: nope, doesn't fix #1106, as seen in the docker build: https://app.circleci.com/pipelines/github/bids-standard/bids-validator/620/workflows/0d7e2bec-2e13-4073-aeee-75163324fc14/jobs/7402

image

### CODE REVIEW

**Comment by sappelhoff** [comment]
@rwblair did you also `cd` into the subdirs and just added `force` to the command? or did you do something else?

Also, this does not seem to fix the deprecation warnings - but perhaps at least the security issues?
(see: https://app.circleci.com/pipelines/github/bids-standard/bids-validator/660/workflows/3eaf4182-90ef-456b-b8ec-568d83ccc82f/jobs/7598 ... and screenshot below)

image



also, Ross did you see: https://github.com/bids-standard/bids-validator/issues/1106#issuecomment-783377429 ?

**Comment by rwblair** [comment]
Yeah ran the audit cd'd into the directory. As far as my lerna/npm concerns as long as the package.json is getting updated it should be good to go.

Looking at #1106.

**Comment by rwblair** [comment]
So my last commit 8c37da4 does quite a few things and we may revert.

The lerna configuration wasn't setup to use yarn, so I removed the yarn.lock and updated documentation to use npm solely.

Running npm install in the root directory should take care of the linking for bids-validator and bids-validator-web. At this point lerna is only being used to manage the semver for package.json files.

Something in the last batch of upgrades broke hed tests so I'll try and fix those.

I'll also update contributing documentation with state of tooling and note npm 7+ as a requirement.

File: package.json [hunk 1]
```diff
@@ -11,10 +11,10 @@
...
     "npmPublish": "cd bids-validator && publish",
     "lernaPublish": "./node_modules/.bin/lerna publish && ./tools/post_publish.sh",
-    "web-dev": "cd bids-validator-web && yarn dev",
-    "web-build": "cd bids-validator-web && yarn build",
-    "web-start": "cd bids-validator-web && yarn build && yarn start",
-    "web-export": "yarn web-build && cd bids-validator-web && yarn next export"
+    "web-dev": "cd bids-validator-web && npm dev",
+    "web-build": "cd bids-validator-web && npm build",
+    "web-start": "cd bids-validator-web && npm build && npm start",
+    "web-export": "npm web-build && cd bids-validator-web && npm next export"
```
Comment (nellh) [inline package.json]: These need to be `npm run ...`  for any scripts being called, the shorthand only works for `npm start`.

**Review by nellh — CHANGES_REQUESTED** [review]
@rwblair This looks pretty good to me. Requiring npm 7 for development seems fine, it should still run on npm 6 or yarn 1.7.0+

File: bids-validator-code.code-workspace [hunk 2]
```diff
@@ -1,8 +0,0 @@
-{
-	"folders": [
-		{
-			"path": "."
-		}
-	],
-	"settings": {}
-}
```
Comment (sappelhoff) [inline bids-validator-code.code-workspace:8]: what was this file for?

File: bids-validator/package.json [hunk 3]
```diff
@@ -11,7 +11,7 @@
     "node": ">=10.11.0"
   },
   "bin": {
-    "bids-validator": "./bin/bids-validator"
+    "bids-validator": "bin/bids-validator"
```
Comment (sappelhoff) [inline bids-validator/package.json:14]: was this an automatic change? or is there some deliberate reasoning to it?

File: bids-validator/package.json [hunk 4]
```diff
@@ -24,14 +24,13 @@
...
     "colors": "^1.3.1",
-    "cross-fetch": "^2.2.2",
+    "cross-fetch": "^3.0.6",
     "date-fns": "^2.7.0",
     "esm": "^3.2.25",
     "hed-validator": "^3.0.0",
     "ignore": "^4.0.2",
     "is-utf8": "^0.2.1",
     "jshint": "^2.9.6",
-    "lerna": "^3.15.0",
```
Comment (sappelhoff) [inline bids-validator/package.json:34]: did you remove this manually? I get confused with what gets auto-updated through things like "npm audit fix", and what gets manually removed.

Also I thought we'd still need lerna for the publishing flow?

explanations would be highly appreciated, so I can be of more help in the future! :)

**Review by sappelhoff — COMMENTED** [review]
so does this now supersede #1102?

File: bids-validator-code.code-workspace (hunk 2 above)
Comment (rwblair) [reply bids-validator-code.code-workspace:8]: Looks like it was a visual studio code specific file that snuck through some how.

File: bids-validator/package.json (hunk 4 above)
Comment (rwblair) [reply bids-validator/package.json:34]: That I removed manually, it was causing a number of the deprecation warnings.

Lerna is only a dev dependency of the root project, not necessary for the bids-validator package that we publish to NPM to work.

File: bids-validator/package.json (hunk 3 above)
Comment (rwblair) [reply bids-validator/package.json:14]: I believe that was automatic, not sure what package.json thinks about relative paths with and without dot slash. If in the end the tests pass on CI and locally I won't worry about it.

**Comment by rwblair** [comment]
> so does this now supersede #1102?

Sorry to hijack your PR, the deprecation warning issue lead me down a rabbit hole, and this PR was newer cleaner than my old branch. I went ahead and closed #1102.

File: bids-validator/package.json (hunk 4 above)
Comment (sappelhoff) [reply bids-validator/package.json:34]: Ahh I see, thanks. perhaps we should also take the chance to bump `"lerna": "^3.15.0"` for 3.22.1 (see https://github.com/bids-standard/bids-validator/issues/1106#issuecomment-783377429)

File: bids-validator/package.json (hunk 3 above)
Comment (sappelhoff) [reply bids-validator/package.json:14]: fair enough, thanks!

**Comment by rwblair** [comment]
~~@happy5214 I'm seeing an invalid place holder error show up in the hed tests, Is this something the bids-validator is doing wrong?~~

Looks like this is generated from hed-validator 3.0.3. For this PR I'll pin to 3.0.2

https://circleci.com/api/v1.1/project/github/bids-standard/bids-validator/7630/output/107/0?file=true&allocation-id=603d73fd600d152d2f0a5cd3-0-build%2F7603ED53

**Comment by happy5214** [comment]
> ~@happy5214 I'm seeing an invalid place holder error show up in the hed tests, Is this something the bids-validator is doing wrong?~
> 
> Looks like this is generated from hed-validator 3.0.3. For this PR I'll pin to 3.0.2
> 
> https://circleci.com/api/v1.1/project/github/bids-standard/bids-validator/7630/output/107/0?file=true&allocation-id=603d73fd600d152d2f0a5cd3-0-build%2F7603ED53

It is in fact an error with the bids-validator testcase for that issue code. The hed-validator code was fixed in that version to only pass placeholder tags (ones with `#`) in tags with the `takesValue` attribute, which that tag does not. The previous behavior (in 3.0.2) was in fact a bug. I can post a PR with an updated test string.

File: bids-validator/package.json [hunk 5]
```diff
@@ -24,14 +24,13 @@
     "aws-sdk": "^2.327.0",
     "bytes": "^3.0.0",
     "colors": "^1.3.1",
-    "cross-fetch": "^2.2.2",
+    "cross-fetch": "^3.0.6",
     "date-fns": "^2.7.0",
     "esm": "^3.2.25",
-    "hed-validator": "^3.0.0",
+    "hed-validator": "3.0.2",
```
Comment (sappelhoff) [inline bids-validator/package.json]: after @happy5214's fix, this can probably be `^3.0.0` again

Comment (sappelhoff) [reply bids-validator/package.json]:
```suggestion
    "hed-validator": "^3.0.0",
```

**Comment by sappelhoff** [comment]
bumping the hed-validator version up again did fix some issues.

The remaining ones are mostly due to:

> Timeout - Async callback was not invoked within the 5000ms timeout specified by jest.setTimeout.Timeout - Async callback was not invoked within the 5000ms timeout specified by jest.setTimeout.Error:

whatever that is.

**Comment by nellh** [comment]
> bumping the hed-validator version up again did fix some issues.
> 
> The remaining ones are mostly due to:
> 
> > Timeout - Async callback was not invoked within the 5000ms timeout specified by jest.setTimeout.Timeout - Async callback was not invoked within the 5000ms timeout specified by jest.setTimeout.Error:
> 
> whatever that is.

That timeout can be two kinds of failures, one is that CI is throttled so these tests finish but they take longer than the 5 second cutoff. The timeout can be raised by passing milliseconds as the third argument to it(). The other is that the promise never resolves and that's a legitimate bug.

**Comment by sappelhoff** [comment]
running `npm test` on a fresh local installation (`npm install` after `rm -rf node_modules`) of this PR passes. :thinking: 

I noticed that in the CI, there is also one non-timeout related error (see screenshot below)  -  but that one passes locally as well.

image

File: .circleci/config.yml [hunk 6]
```diff
@@ -33,23 +33,24 @@ jobs:
...
           name: Eslint
-          command: yarn lint
+          command: npm run lint
       - run:
           name: Get bids-examples data
           command: git submodule update --init
       - run:
           name: Jest tests
-          command: yarn coverage --maxWorkers=2
+          command: npm run coverage --maxWorkers=2
```
Comment (sappelhoff) [inline .circleci/config.yml]:
```suggestion
          command: npm run coverage --maxWorkers=2 --testTimeout= 10000
```

**Comment by sappelhoff** [comment]
@nellh @rwblair the commit I made in https://github.com/bids-standard/bids-validator/pull/1187/commits/fbd336005da37302b08675fd36e1ce7255e19d5b fixes the test issues ... so the problem is not with the code or data per se, but in the way the tests are triggered (npm run, etc.) ... now it fails with the codecov command (the step afterwards).

Not sure what needs fixing here.

**Comment by sappelhoff** [comment]
>  so the problem is not with the code or data per se, but in the way the tests are triggered (npm run, etc.) ...

and the real issue was, that when trying to pass args to an "npm script", one needs to do it this way: `npm run <scriptname> -- --<arg1> --<arg2>` ... without the separating `--`, the args are silently passed to (and swallowed by) npm instead of to `<scriptname>`

the codecov job was probably an issue because it was not part of dev dependencies in the root package.json. but we'll see.

**Comment by sappelhoff** [comment]
okay, I tried several things to make the githubPagesTest work ... I am giving up for now. The main issue is that the docker image used for that test was not NPM v7 ... so the last 4 commits can probably be reverted and replaced by an appropriate commit that results in npm v7 for that job.

**Comment by sappelhoff** [comment]
one step further: githubPagesTest is now running, but breaking during the web-validator install:

> ModuleNotFoundError: Module not found: Error: Can't resolve 'pluralize' in '/root/project/bids-validator-web/components'

**Comment by sappelhoff** [comment]
this is an absolute rabbit hole: Fix one issue, get the next one. :)

> 
> Build error occurred
Error: > Build failed because of webpack errors
    at /root/project/node_modules/next/dist/build/index.js:15:918
    at async /root/project/node_modules/next/dist/build/tracer.js:3:470

**Comment by sappelhoff** [comment]
@happy5214, there was a test failure with hed-validator `3.0.5` that was resolved when downgrading to `3.0.4` (in https://github.com/bids-standard/bids-validator/pull/1187/commits/a395fd2e7de56a191682da34b997b5d2a4724b30). I only stumbled across this now, perhaps because my commit https://github.com/bids-standard/bids-validator/pull/1187/commits/d2d6a669490dfcd7f45f87adc5f4f89664e57b6a triggered the cache to be cleared.

Could you investigate, please?

**Pull request merged**